design/authenticators/authenticator_refactor.md

```diff
@@ -0,0 +1,284 @@
+        client_id
+        client_secret
+        redirect_uri
+        claim_mapping
+      ].freeze
+
+      REQUIRED_PARAMS = %i[code state].freeze
+
+      def initialize(config:)
+        # Validation will be moved upstream
+        validate_present(
+          args: config,
+          required_fields: REQUIRED_CONFIGURATION
+        )
+        @config = config
+        # ...
+      end
+
+      def call(params:)
+        # Validation will be moved upstream
+        validate_present(
+          args: params,
+          required_fields: REQUIRED_PARAMS
+        )
+        # Identify the OIDC user resulting from the callback
+      end
+      # ...
+    end
+  end
+end
+```
+
+### OIDC Mapper
+
+Given the allowed Conjur Roles, the mapper translates selects the Role to bind the external identity to:
+
+```ruby
+module Authentication
+  module AuthnOidc
+    class Mapper
+      def initialize(allowable_roles:)
+        @allowable_roles = allowable_roles
+      end
+
+      def call(identity:)
+        identified_role = @allowable_roles.select do |role|
+          role.id.split(':').last == identity
+        end
+        raise "#{identify} was not found as a valid authenticatable role" if identified_role.nil?
+
+        identified_role
+      end
+    end
+  end
+end
+```
+
+### Authenticator Workflow
+
+Finally, we glue the above bits together:
+
+```ruby
+module Validations
+  class Cidr
+    def valid?(role:, ip:)
+      raise Errors::Authentication::InvalidOrigin unless role.valid_origin?(ip)
+    end
+  end
+end
+
+class Authenticator
+  def initialize(token_factory: ::TokenFactory)
+    @token_factory = token_factory
+  end
+
+  def call(account:, request:, identity_resolver:, mapper:)
+    role = mapper.call(
+      identity: identity_resolver.call(
+        params: request
+      )
+    )
+    Validations::Cidr.new.valid?(role: role, ip: request[:ip])
+
+    @token_factory.signed_token(
+      account: account,
+      username: role
+    )
+  end
+end
+```
+
+### Full Workflow
+
+With these parts, we can construct a full authenticator flow with the following:
+
+```ruby
+authenticator_configuration = Authenticators::AuthenticatorLoader.new.find(
+  account: params[:account],
+  service_id: params[:service_id],
+  type: 'authn-oidc'
+)
+
+Authenticator.new.call(
+  identity_resolver: Authentication::AuthnOidc2::IdentityResolver.new(
+    config: authenticator_configuration.config
+  ),
+  mapper: Authentication::AuthnOidc2::Mapper.new(
+    roles: authenticator_configuration.authenticatable_roles
+  ),
+  params: params,
+  account: params[:account]
+)
+```
+
+The result of the `Authenticator` call is an Conjur Authorization Token, which is returned to the newly authorized user.
+
+## Advantages
+
+- Most authenticators (`authn-azure`, `authn-gcp`, `authn-jwt`, `authn-oidc`) only perform JWT validation and attribute extraction. With the above pattern we can write a single JWT validation class while providing custom JWT mapping.
```

Comment:
This also increases security by making sure that all authenticators handle the various ways that JWT validation can be gotten wrong.